### PR TITLE
[JAX] Support segment_ids/pos as FA inputs

### PR DESCRIPTION
**Pull request by zlsh80826**
# Description

This PR adds `segment_ids/pos` limited support and deprecated `fused_attn_thd` API.

## Type of change

- [x] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

- Add a new `SequenceDescriptor` class for different sequence descriptions scenario.
   - `from_seqlens` for non-THD
   - `from_seqlens_and_offsets` for THD
   - `from_segment_ids_and_pos` for THD + ring attn (haven't implemented)
- Change the old `fused_attn` `mask` parameter to `SequenceDescriptor`. Passing `mask` in the position argument will work for a while but generating deprecation warning.
- Deprecate `fused_attn_thd` API as the refactored `fused_attn` can also support THD format.
- Remove small inputs in `test_fused_attn.py` as the long sequence inputs should cover.
- Add different sequence inputs tests in `test_fused_attn.py`

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/jax/attention.py [hunk 1]
```diff
@@ -289,7 +288,270 @@ def inverse_reorder_causal_load_balancing(tensor, cp_size: int, tensor_format: Q
...
+            A SequenceDescriptor with segment_ids/segment_pos initialized.
+        """
+        q_seg_ids, kv_seg_ids = cls._expand_to_pair(segment_ids)
+
+        if segment_pos is not None:
+            segment_pos = cls._expand_to_pair(segment_pos)
+        else:
+
+            def generate_default_pos(segment_ids):
+                seqlen = segment_ids.shape[-1]
```
[the remark this answers is not in the document]
Comment (zlsh80826) [reply transformer_engine/jax/attention.py:545]: I added https://github.com/NVIDIA/TransformerEngine/pull/1406/commits/fea6b0e0e96a76a53f48d0da003dcc78e53ef792 and https://github.com/NVIDIA/TransformerEngine/pull/1406/commits/cfe00a89a71e6ca012ad4c46897aab8fd1df8942 to reduce the `seqlen` and `offset` shape to `(batch, max_segments_per_seq)`

File: tests/jax/test_fused_attn.py [hunk 2]
```diff
@@ -709,10 +727,7 @@ def check_dqkv(primitive, reference, pad):
 @pytest.mark.parametrize(
     "b, s_q, s_kv, h_q, h_kv, d, dtype",
     [
-        pytest.param(4, 128, 128, 16, 16, 64, jnp.bfloat16, id="4-128-128-16-16-64-BF16-SELF"),
```
[the remark this answers is not in the document]
Comment (zlsh80826) [reply tests/jax/test_fused_attn.py:712]: Yes, I think those tests are unnecessary because we no longer use [max_512 kernels](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/common/fused_attn/fused_attn_f16_max512_seqlen.cu) for seqlen < 512. The kernels for seqlen=128 are the same as seqlen=2048, so the 2048 test cases already cover the 128 cases.

File: tests/jax/test_fused_attn.py [hunk 3]
```diff
@@ -741,6 +756,14 @@ def check_dqkv(primitive, reference, pad):
         pytest.param(True, id="SWA"),
     ],
 )
+@pytest.mark.parametrize(
+    "seq_desc_format",
+    [
+        pytest.param(SeqDescFormat.Mask, id="Mask"),
```
[the remark this answers is not in the document]
Comment (zlsh80826) [reply tests/jax/test_fused_attn.py:762]: Good idea

**Review by mgoldfarb-nvidia — APPROVED** [review]
LGTM.

File: tests/jax/test_fused_attn.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zlsh80826) [reply tests/jax/test_fused_attn.py:939]: I think we still need to keep those different `SeqDescFormat` in the `test_fused_attn` for a while. I don't have a better way to verify the converted seqlens/offsets are correct, the best way to verify that is checking the FA results. The current time for CI is still under 1hr so I think it is ok for now. But yeap, if we found some of SeqDescFormat are rarely used in the future, we can remove them or just testing on the specific input shape instead of all input shapes.

**Review by phu0ngng — APPROVED** [review]
